### PR TITLE
C++: match extractor changes to two-operand ?: representation

### PR DESCRIPTION
**Pull request by nickrolfe**
Closes https://github.com/github/codeql-c-extractor-team/issues/67

Now the extractor emits a synthesized expr node for the 'then' case.

I've updated the `question_mark_colon` test to show that the 'then' expr now includes any implicit cast.

The upgrade script populates the missing `expr_cond_true` tuples by re-using the guard expr (just like the old implementation of `getThen()`).

I could use some assistance in judging whether the new results for the consistency queries are a problem.

### CODE REVIEW

**Comment by rdmarsh2** [comment]
I think both the consistency query results and the simple range analysis changes will need library fixes.

**Comment by nickrolfe** [comment]
I think there is an extractor issue for cases where there *isn't* an implicit cast on the 'then' expr. I'm investigating.

**Comment by nickrolfe** [comment]
That extractor issue has been fixed, but the consistency query and simple range analysis results haven't changed. I'm not sure what kind of library changes are required there.

File: cpp/ql/test/library-tests/ir/ir/PrintAST.expected [hunk 1]
```diff
@@ -10449,7 +10479,37 @@ ir.cpp:
...
 # 1308|                     ValueCategory = prvalue(load)
-# 1308|           1: [VariableAccess] x
+# 1308|           1: [CStyleCast] (int)...
+# 1308|               Conversion = [IntegralConversion] integral conversion
+# 1308|               Type = [IntType] int
+# 1308|               ValueCategory = prvalue
+# 1308|             expr: [ParenthesisExpr] (...)
+# 1308|                 Type = [BoolType] bool
+# 1308|                 ValueCategory = prvalue
+# 1308|               expr: [LogicalOrExpr] ... || ...
```
Comment (dbartol) [inline cpp/ql/test/library-tests/ir/ir/PrintAST.expected:10489]: Is this the exact same expression object as child 0, or is it a copy? What happens if operand 0 might have side effects, as in `foo() ?: 7`?

**Comment by dbartol** [comment]
@jbj Are you OK with us merging this PR (and the extractor change) now, with an issue opened for me to go back and fix the IR consistency failures? Or should I try to contribute the necessary IR construction fixes to this PR before we merge it?

File: cpp/ql/test/library-tests/ir/ir/PrintAST.expected (hunk 1 above)
Comment (nickrolfe) [reply cpp/ql/test/library-tests/ir/ir/PrintAST.expected:10489]: In this case, this is a separate entity in the db, but getting the same entity is possible, such as in your `foo()` example.

You can see an example of an expr being reused like that in `question_mark_colon/types.expected` that I added in this PR.

**Comment by nickrolfe** [comment]
I made this change in response to the extractor issue you raised, so if it's not a net-positive for IR consistency, I have no issue with postponing the merge.

**Comment by dbartol** [comment]
I believe (but have not confirmed) that your change to fix the extractor issue merely breaks the workaround that I was using before. I just have to go back into the IR construction code and change it do do the right thing, rather than the workaround.

**Comment by jbj** [comment]
Ping @dbartol.

**Pull request closed without merging**